File: pp.pl
pp(N, L) :-
    findall(X, primes(1, N, X), [H|_]),
    subseq(L, H),
    sum_list(L, N).

subseq([A|B], [A|D]) :- subseq(B, D).
subseq([A|B], [_|D]) :- subseq([A|B], D).
subseq([], _).

sum_list([],0).
sum_list([A|B], N) :-
    Y is N - A,
    sum_list(B, Y).

primes(A, B, AllPrimes) :-
    numlist(A, B, List),
    include(isPrime, List, AllPrimes).

isPrime(N) :-
    \+ (divisible(N, N - 1)).

divisible(N, K) :-
    N < 2 ; (K > 1, (mod(N, K) =:= 0 ; divisible(N, K - 1))).

File: pp.ml
let rec divisible n k =
        ((k > 1) && ((n mod k == 0) || divisible n (k-1))) || (n < 2);;

let isP n = 
        divisible n (n-1);;



let rec primes a b = 
        if a > b then [] else
                let tail = primes (a+1) b in
                if not (isP a) then a :: tail else tail;;

let rec primePartition n k =
        List.flatten (List.map (fun i -> if i == n then [[i]] else 
                (List.map (fun y -> i::y) 
        (primePartition (n-i) (i+1)))) (primes k n));;
                

let print_list ls =
        print_string(String.concat " + " (List.map string_of_int ls));
        print_string "\n";;


let () = print_string "type a number: ";;
let i = read_int();;


let xd = primePartition i 1 ;;

List.iter (print_list) xd;;
